connectors/sources/salesforce.py

```diff
@@ -0,0 +1,1443 @@
+            emails.append(participant.get("Email"))
+            names.append(participant.get("Name"))
+
+        ids = self._format_list(ids)
+        emails = self._format_list(emails)
+        names = self._format_list(names)
+
+        return ids, emails, names
+
+    def _collect_created_by(self, sobject):
+        created_by_list = []
+
+        records = sobject.get("records", [])
+        for record in records:
+            created_by_list.append(record.get("CreatedBy"))
+
+        return created_by_list
+
+    def _format_list(self, unformatted):
+        return sorted(
+            set(filter(None, [str(x).strip() for x in unformatted if x is not None]))
+        )
+
+
+class SalesforceDataSource(BaseDataSource):
+    """Salesforce"""
+
+    name = "Salesforce"
+    service_type = "salesforce"
+
+    def __init__(self, configuration):
+        super().__init__(configuration=configuration)
+        self.base_url = BASE_URL.replace("<domain>", configuration["domain"])
+        self.salesforce_client = SalesforceClient(
+            configuration=configuration, base_url=self.base_url
+        )
+        self.doc_mapper = SalesforceDocMapper(self.base_url)
+
+    def _set_internal_logger(self):
+        self.salesforce_client.set_logger(self._logger)
+
+    @classmethod
+    def get_default_configuration(cls):
+        return {
+            "client_id": {
+                "label": "Client ID",
+                "order": 1,
+                "tooltip": "The client id for your OAuth2-enabled connected app. Also called 'consumer key'",
+                "type": "str",
+                "value": "",
+            },
+            "client_secret": {
+                "label": "Client Secret",
+                "order": 2,
+                "sensitive": True,
+                "tooltip": "The client secret for your OAuth2-enabled connected app. Also called 'consumer secret'",
+                "type": "str",
+                "value": "",
+            },
+            "domain": {
+                "label": "Domain",
+                "order": 3,
+                "sensitive": True,
+                "tooltip": "The domain for your Salesforce instance. If your Salesforce URL is 'foo.my.salesforce.com', the domain would be 'foo'.",
+                "type": "str",
+                "value": "",
+            },
+        }
+
+    async def validate_config(self):
+        await super().validate_config()
+
+    async def close(self):
+        await self.salesforce_client.close()
+
+    async def ping(self):
+        try:
+            await self.salesforce_client.ping()
+            self._logger.debug("Successfully connected to Salesforce.")
+        except Exception as e:
+            self._logger.exception(f"Error while connecting to Salesforce: {e}")
+            raise
+
+    async def get_docs(self, filtering=None):
+        # We collect all content documents and de-duplicate them before downloading and yielding
+        all_content_docs = []
```

Comment:
> We collect all content documents and de-duplicate them before downloading and yielding

What does collecting all content documents mean exactly here? Is this expected to be a lot for large datasets? Can we cause OOMs with that reference?

Reply:
Every SObject can have a x number of related ContentDocuments, but the same ContentDocument can be attached to multiple SObjects. This collection is so we can compare them before downloading and avoid downloading duplicate files.

> Can we cause OOMs with that reference?
It's possible. This is something we will find out in QA when we can test against a very large dataset. We can exchange the memory usage for higher CPU usage by checking duplicates on the fly, but we deemed it wasn't necessary because 1. we need to see if it's actually a problem, and 2. it's still a beta connector so it wasn't urgent to fix.

[Discussion link](https://github.com/elastic/connectors-python/pull/1451#discussion_r1294483719)